fiber/chain_interactions/commitments.py

```diff
@@ -0,0 +1,224 @@
+from enum import Enum
+from typing import TypeAlias
+
+from pydantic import BaseModel
+from scalecodec import ScaleType
+from substrateinterface import SubstrateInterface, Keypair
+from tenacity import retry, stop_after_attempt, wait_exponential
+
+from fiber.chain_interactions.chain_utils import format_error_message
+
+
+class DataFieldType(Enum):
+    RAW = "Raw"
+    BLAKE_TWO_256 = "BlakeTwo256"
+    SHA_256 = "Sha256"
+    KECCAK_256 = "Keccak256"
+    SHA_THREE_256 = "ShaThree256"
+
+
+CommitmentDataField: TypeAlias = tuple[DataFieldType, bytes] | None
+
+
+class CommitmentQuery(BaseModel):
+    fields: list[CommitmentDataField]
+    block: int
+    deposit: int
+
+
+class RawCommitment(BaseModel):
+    data: bytes
+    block: int
+    deposit: int
+
+
+def _serialize_field(field: CommitmentDataField):
+    if not field:
+        return {str(None): b''}
+
+    data_type, data = field
+
+    if data_type == DataFieldType.RAW:
+        serialized_data_type = DataFieldType.RAW.value + str(len(data))
+    else:
+        serialized_data_type = data_type.value
+
+    return {serialized_data_type: data}
+
+
+def _deserialize_field(field: dict[str, bytes | str]) -> CommitmentDataField:
+    data_type, data = field.items().__iter__().__next__()
+
+    if data_type == str(None):
+        return None
+
+    if data_type.startswith(DataFieldType.RAW.value):
+        expected_length = int(data_type[len(DataFieldType.RAW.value):])
+        data_type = DataFieldType.RAW.value
+        data = bytes.fromhex(data[2:])
+
+        if len(data) != expected_length:
+            raise ValueError(f"Got commitment raw field expecting {expected_length} data but got {len(data)} data")
+
+    return DataFieldType(data_type), data
+
+
+@retry(
+    stop=stop_after_attempt(3),
+    wait=wait_exponential(multiplier=1, min=1, max=4),
+    reraise=True,
+)
+def _query_commitment(
+    substrate: SubstrateInterface,
+    netuid: int,
+    hotkey: str,
+    block: int | None = None,
+) -> ScaleType:
+    return substrate.query(
```

Comment:
substrate is better than substrate_interface though, i will replace it to that everywhere thanks.

Note for this, you may need to close the connection after or reinstantiate if it fails. I'm going to try to make a fixed SubstrateInterface class for this, which handles the constant SSL errors we're seeing

Reply:
so basically that was a note to self ^